cis194/week6/daniel-mc/Fibonacci.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -Wall #-}
+
+module Fibonacci
+  ( fib
+  , fibs1
+  , fibs2
+  , streamToList
+  , streamRepeat
+  , streamMap
+  , streamFromSeed
+  , nats
+  , ruler
+  , fibs3
+  , fibs4
+  )
+where
+
+-- Exercise 1
+
+fib :: Integer -> Integer
+fib 0 = 0
+fib 1 = 1
+fib n = fib (n - 1) + fib (n - 2)
+
+fibs1 :: [Integer]
+fibs1 = map fib [0 ..]
+
+
+-- Exercise 2
+iterate2 :: (a -> a -> a) -> a -> a -> [a]
+iterate2 f x y = x : y : iterate2 f x' (f y x') where x' = f x y
+
+fibs2 :: [Integer]
+fibs2 = iterate2 (+) 0 1
+
+-- Exercise 3
+data Stream a = Stream a (Stream a)
+
+streamToList :: Stream a -> [a]
+streamToList (Stream x xs) = x : streamToList xs
+
+instance Show a => Show (Stream a) where
+  show = show . take 20 . streamToList
+
+-- Exercise 4
+streamRepeat :: a -> Stream a
+streamRepeat x = Stream x (streamRepeat x)
+
+streamMap :: (a -> b) -> Stream a -> Stream b
+streamMap f (Stream x xs) = Stream (f x) (streamMap f xs)
```

Comment:
Not sure which one's better. Is it a subjective thing? @shineli1984 

```suggestion
streamMap f (Stream x xs) = Stream (f x) $ streamMap f xs
```

Reply:
Not sure but I think we prefer $

Reply:
The general rule is that we want things gluing functions and/or variables together to be as succinct/invisible as possible. So that we can express ourselves by building bigger pieces from tiny ones (or even almost nothing) without introduce too much visual distraction. That's why `.` is a perfect fit as a composition operator. 
Here, `$` is preferred following the same rule, because `(`,`)` are more visible than a single character `$`.